### PR TITLE
Add Timeout to Regex used in the tokenizers

### PR DESCRIPTION
**Pull request by tarekgh**

### CODE REVIEW

File: src/Microsoft.ML.Tokenizers/Model/SentencePieceTokenizer.cs [hunk 1]
```diff
@@ -86,6 +86,8 @@ private SentencePieceTokenizer(ModelProto modelProto, IReadOnlyDictionary<string
                     _specialTokensReverse.Add(item.Value, item.Key);
                 }
 
+                // We create this Regex object without a timeout because we know it will perform the match operation in O(N) time complexity.
+                // Therefore, confidently process any text within a reasonable time.
```
Comment (stephentoub) [inline src/Microsoft.ML.Tokenizers/Model/SentencePieceTokenizer.cs:90]: I'd remove this last sentence. The concern isn't the total time required but whether a small amount of input data could result in a disproportionate amount of conputation required.

File: src/Microsoft.ML.Tokenizers/Model/TiktokenTokenizer.cs [hunk 2]
```diff
@@ -1185,13 +1185,13 @@ private static (Dictionary<string, int> SpecialTokens, Regex Regex, string Vocab
         internal static partial Regex O200kBaseRegex();
 #else
         private static Regex? _cl100kBaseRegex;
-        private static Regex Cl100kBaseRegex() => _cl100kBaseRegex ??= new Regex(Cl100kBaseRegexPattern, RegexOptions.Compiled);
+        private static Regex Cl100kBaseRegex() => _cl100kBaseRegex ??= new Regex(Cl100kBaseRegexPattern, RegexOptions.Compiled, TimeSpan.FromMilliseconds(PreTokenizer.DefaultTimeOutInMilliseconds));
```
Comment (stephentoub) [inline src/Microsoft.ML.Tokenizers/Model/TiktokenTokenizer.cs:1188]: Don't you also need to update the corresponding [GeneratedRegex(...)]?

File: src/Microsoft.ML.Tokenizers/PreTokenizer/RegexPreTokenizer.cs [hunk 3]
```diff
@@ -35,6 +35,8 @@ public RegexPreTokenizer(Regex regex, IReadOnlyDictionary<string, int>? specialT
 
             if (specialTokensEncoder is { Count: > 0 })
             {
+                // We create this Regex object without a timeout because we know it will perform the match operation in O(N) time complexity.
+                // Therefore, confidently process any text within a reasonable time.
```
Comment (stephentoub) [inline src/Microsoft.ML.Tokenizers/PreTokenizer/RegexPreTokenizer.cs:39]: same here

File: src/Microsoft.ML.Tokenizers/PreTokenizer/RegexPreTokenizer.cs [hunk 4]
```diff
@@ -35,6 +35,8 @@ public RegexPreTokenizer(Regex regex, IReadOnlyDictionary<string, int>? specialT
 
             if (specialTokensEncoder is { Count: > 0 })
             {
+                // We create this Regex object without a timeout because we know it will perform the match operation in O(N) time complexity.
```
Comment (stephentoub) [inline src/Microsoft.ML.Tokenizers/PreTokenizer/RegexPreTokenizer.cs:38]: It's actually O(N * NumberOfKeys), but if the latter is a constant, then it's still O(N).

File: src/Microsoft.ML.Tokenizers/Model/SentencePieceTokenizer.cs (hunk 1 above)
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Model/SentencePieceTokenizer.cs:90]: I'll remove the sentence.

File: src/Microsoft.ML.Tokenizers/Model/TiktokenTokenizer.cs (hunk 2 above)
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Model/TiktokenTokenizer.cs:1188]: Yes, I missed that. Thanks for the catch.

File: src/Microsoft.ML.Tokenizers/PreTokenizer/RegexPreTokenizer.cs (hunk 4 above)
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/PreTokenizer/RegexPreTokenizer.cs]: Yes, NumberOfKeys is constant after creating the pre-tokenizer. I added extra sentence clarifying that.

**Review by stephentoub — APPROVED** [review]

**Comment by tarekgh** [comment]
Thanks @stephentoub for your feedback and review!

**Pull request merged**